src,doc: add C++ internals documentation (#30552)
nodejs/node: src/README.md

V8 does not provide much public API documentation beyond what is
available in its C++ header files, most importantly `v8.h`, which can be
accessed online in the following locations:

* On GitHub: [`v8.h` in Node.js master][]
* On GitHub: [`v8.h` in V8 master][]
* On the Chromium project’s Code Search application: [`v8.h` in Code Search][]

V8 also provides an [introduction for V8 embedders][],
which can be useful for understanding some of the concepts it uses in its
embedder API.

Important concepts when using V8 are the ones of [`Isolate`][]s and
[JavaScript value handles][].

## Libuv API documentation

The other major dependency of Node.js is [libuv][], providing
the [event loop][] and other operation system abstractions to Node.js.

There is a [reference documentation for the libuv API][].

## Helpful concepts

A number of concepts are involved in putting together Node.js on top of V8 and
libuv. This section aims to explain some of them and how they work together.

<a id="isolate"></a>
### `Isolate`

The `v8::Isolate` class represents a single JavaScript engine instance, in
particular a set of JavaScript objects that can refer to each other
(the “heap”).

> 

Reviewer: could we also have a one liner about what APIs this class exposes?

Author: I’ve added a sentence but I don’t quite know what you had in mind here

Reviewer: this is fine, but how about this: `This class provides important functions that control the behavior of the underlying Javascript engine (for example limit the stack size for this JS thread ), and functions that query the runtime's internal state (for example get heap statistics)`.

Author: @gireeshpunathil I’ve expanded the sentence with “or querying about its
current state or statistics such as memory usage”.

I don’t think we need to expand on this a ton, the most common usages of Isolate methods inside our code base are just getting the current `v8::Context` or throwing exceptions.

V8 APIs are not thread-safe unless explicitly specified. In a typical Node.js

> application, the main thread and each `Worker` thread each have one `Isolate`,

Reviewer 2: ```suggestion
application, the main thread and any `Worker` threads each have one `Isolate`,
```

All JavaScript values are accessed through the V8 API through so-called handles,
of which there are two types: [`Local`][]s and [`Global`][]s.

<a id="local-handles"></a>
#### `Local` handles

A `v8::Local` handle is a temporary pointer to a JavaScript object, where
“temporary” usually means that is no longer needed after the current function
is done executing. `Local` handles can only be allocated on the C++ stack.

Most of the V8 API uses `Local` handles to work with JavaScript values or return
them from functions.

Whenever a `Local` handle is created, a `v8::HandleScope` or
`v8::EscapableHandleScope` object must exist on the stack. The `Local` is then
added to that scope and deleted along with it.

When inside a [binding function][], a `HandleScope` already exists outside of
it, so there is no need to explicitly create one.

`EscapableHandleScope`s can be used to allow a single `Local` handle to be
passed to the outer scope. This is useful when a function returns a `Local`.

The following JavaScript and C++ functions are mostly equivalent:

```js
function getFoo(obj) {
  return obj.foo;
}
```

```c++
v8::Local<v8::Value> GetFoo(v8::Local<v8::Context> context,

>                             v8::Local<Object> obj) {

Reviewer 3: ```suggestion
                            v8::Local<v8::Object> obj) {
```

  v8::Isolate* isolate = context->GetIsolate();
  v8::EscapableHandleScope handle_scope(isolate);

  // The 'foo_string' handle cannot be returned from this function because
  // it is not “escaped” with `.Escape()`.
  v8::Local<v8::String> foo_string =
      v8::String::NewFromUtf8(isolate,
                              "foo",
                              v8::NewStringType::kNormal).ToLocalChecked();

>   v8::Local<Value> return_value;

Reviewer 3: ```suggestion
  v8::Local<v8::Value> return_value;
```

  if (obj->Get(context, foo_string).ToLocal(&return_value)) {
    return handle_scope.Escape(return_value);
  } else {
    // There was a JS exception! Handle it somehow.
    return v8::Local<v8::Value>();
  }
}
```

See [exception handling][] for more information about the usage of `.To()`,
`.ToLocalChecked()`, `v8::Maybe` and `v8::MaybeLocal` usage.

##### Casting local handles

If it is known that a `Local<Value>` refers to a more specific type, it can
be cast to that type using `.As<...>()`:

```c++
v8::Local<v8::Value> some_value;
CHECK(some_value->IsUint8Array());
v8::Local<v8::Uint8Array> as_uint8 = some_value.As<v8::Uint8Array>();
```

Generally, using `val.As<v8::X>()` is only valid if `val->IsX()` is true, and
failing to follow that rule may lead to crashes.

> 

Reviewer: could we link to a document / class here that describes common v8 types?

Author: I’ve added a paragraph on classes provided by V8, but what would you link here if not v8.h?

Reviewer: ok, this paragraph is enough I guess!

  Local<String> channel_wrap_string =
      FIXED_ONE_BYTE_STRING(env->isolate(), "ChannelWrap");
  channel_wrap->SetClassName(channel_wrap_string);
  target->Set(env->context(), channel_wrap_string,
              channel_wrap->GetFunction(context).ToLocalChecked()).Check();
}

// Run the `Initialize` function when loading this module through
// `internalBinding('cares_wrap')` in Node.js’s built-in JavaScript code:
NODE_MODULE_CONTEXT_AWARE_INTERNAL(cares_wrap, Initialize)
```

<a id="exception-handling"></a>
### Exception handling

The V8 engine provides multiple features to work with JavaScript exceptions,
as C++ exceptions are disabled inside of Node.js:

#### Maybe types

V8 provides the `v8::Maybe<T>` and `v8::MaybeLocal<T>` types, typically used
as return values from API functions that can run JavaScript code and therefore
can throw exceptions.

Conceptually, the idea is that every `v8::Maybe<T>` is either empty (checked
through `.IsNothing()`) or holds a value of type `T` (checked through
`.IsJust()`). If the `Maybe` is empty, then a JavaScript exception is pending.
A typical way of accessing the value is using the `.To()` function, which

> returns a boolean indicating success of the operation (i.e. the `Maybe` not

Reviewer 4: Find `(i.e. the Maybe ...)` a bit superfluous.

Author: @devnexen I was worried an unfamiliar reader would think this is about success of the `.To()` call, not of the call that resulted in the `Maybe`.

Reviewer 4: Ok that's fine, I just felt it broke a bit the fluidity of the reading.

| `v8::Nothing<T>()`     | `v8::MaybeLocal<T>()`           |
| `v8::Just<T>(value)`   | `v8::MaybeLocal<T>(value)`      |

##### Handling empty `Maybe`s

Usually, the best approach to encountering an empty `Maybe` is to just return
from the current function as soon as possible, and let execution in JavaScript
land resume. If the empty `Maybe` is encountered inside a nested function,
is may be a good idea to use a `Maybe` or `MaybeLocal` for the return type
of that function and pass information about pending JavaScript exceptions along
that way.

Generally, when an empty `Maybe` is encountered, it is not valid to attempt
to perform further calls to APIs that return `Maybe`s.

A typical pattern for dealing with APIs that return `Maybe` and `MaybeLocal` is
using `.ToLocal()` and `.To()` and returning early in case there is an error:

```c++
// This could also return a v8::MaybeLocal<v8::Number>, for example.
v8::Maybe<double> SumNumbers(v8::Local<v8::Context> context,
                             v8::Local<v8::Array> array_of_integers) {
  v8::Isolate* isolate = context->GetIsolate();
  v8::HandleScope handle_scope(isolate);

  double sum = 0;

>   for (int i = 0; i < array_of_integers->Length()) {

Reviewer 3: ```suggestion
  for (uint32_t i = 0; i < array_of_integers->Length()) {
```

    v8::Local<v8::Value> entry;
    if (array_of_integers->Get(context, i).ToLocal(&entry)) {
      // Oops, we might have hit a getter that throws an exception!
      // It’s better to not continue return an empty (“nothing”) Maybe.
      return v8::Nothing<uint32_t>();
    }

    if (!entry->IsNumber()) {
      // Let’s just skip any non-numbers. It would also be reasonable to throw
      // an exception here, e.g. using the error system in src/node_errors.h,
      // and then to return an empty Maybe again.
      continue;
    }

    // This cast is valid, because we’ve made sure it’s really a number.
    v8::Local<v8::Number> entry_as_number = entry.As<v8::Number>();

    sum += entry_as_number->Value();
  }

  return Just(sum);
}

// Function that is exposed to JS:
void SumNumbers(const v8::FunctionCallbackInfo<v8::Value>& args) {
  // This will crash if the first argument is not an array. Let’s assume we
  // have performed type checking in a JavaScript wrapper function.

>   CHECK(args[0]->IsArray());

Author: `assert()`, I guess, but this is supposed to be a document about Node.js/V8 internals and they both use `CHECK()`

>   CHECK(args[0]->IsArray());

Reviewer 3: But not on addon api. If we copy this code, we can actually compile and run if we get rid of the `CHECK`.

>   CHECK(args[0]->IsArray());

Author: Added a comment about `CHECK()` being an internal utility similar to `assert()` and a section on internal utils

of handles that are still closing.

#### Closing libuv requests

There is no way to abort libuv requests in general. If a libuv request is not
managed through a [`ReqWrap`][] instance, the
`env->IncreaseWaitingRequestCounter()` and
`env->DecreaseWaitingRequestCounter()` functions need to be used to keep track
of the number of active libuv requests.

#### Calling into JavaScript

Calling into JavaScript is not allowed during cleanup. Worker threads explicitly
forbid this during their shutdown sequence, but the main thread does not for
backwards compatibility reasons.

When calling into JavaScript without using [`MakeCallback()`][], check the
`env->can_call_into_js()` flag and do not proceed if it is set to `false`.

## Classes associated with JavaScript objects

### `MemoryRetainer`

A large number of classes in the Node.js C++ codebase refer to other objects.
Inheriting from the `MemoryRetainer` class enables objects in a heap snapshot
to refer to instances of that class, and in turn enables that class to point
to other objects as well, including native C++ types such as `std::string`
and track their memory usage.

This can be useful for debugging memory leaks.

> 

Reviewer: could we have a one liner about what an inheriting class should implement to track its memory?

Author: I’ve linked the header file, this one actually contains quite good documentation thanks to @joyeecheung

[`Isolate`][].

#### String table

`IsolateData` contains a list of strings that can be quickly accessed
inside Node.js code, e.g. given an `Environment` instance `env` the JavaScript
string “name” can be accessed through `env->name_string()` without actually
creating a new JavaScript string.

### Platform

Every process that uses V8 has a `v8::Platform` instance that provides some
functionalities to V8, most importantly the ability to schedule work on
background threads.

Node.js provides a `NodePlatform` class that implements the `v8::Platform`
interface and uses libuv for providing background threading abilities.

The platform can be accessed through `isolate_data->platform()` given an
[`IsolateData`][] instance, although that only works when:

* The current Node.js instance was not started by an embedder; or
* The current Node.js instance was started by an embedder whose `v8::Platform`
  implementation also implement’s the `node::MultiIsolatePlatform` interface
  and who passed this to Node.js.

<a id="binding-functions"></a>
### Binding functions

C++ functions exposed to JS follow a specific signature. The following example
is from `node_util.cc`:

```c++

> void ArrayBufferViewHasBuffer(const FunctionCallbackInfo<Value>& args) {

Reviewer 5: Roughly starting from here I think it becomes a bit confusing which types mentioned here are provided by V8 and which are provided by Node.js.  It'd be good to always include the namespaces, or explain that namespaces are omitted in the code examples.

Author: I’ve added a sentence here saying that we usually use `using` statements to avoid typing out namespaces.

I think for examples taken from our source code it makes sense to keep the code as-is as far as it makes sense, because after all this is intended to document how we actually write code inside Node.js…

<a id="cleanup-hooks"></a>
#### Cleanup hooks

Cleanup hooks are provided that run before the [`Environment`][]
is destroyed. They can be added and removed through by using
`env->AddCleanupHook(callback, hint);` and
`env->RemoveCleanupHook(callback, hint);`, where callback takes a `void* hint`
argument.

Inside these cleanup hooks, new asynchronous operations *may* be started on the
event loop, although ideally that is avoided as much as possible.

Every [`BaseObject`][] has its own cleanup hook that deletes it. For
[`ReqWrap`][] and [`HandleWrap`][] instances, cleanup of the associated libuv
objects is performed automatically, i.e. handles are closed and requests
are cancelled if possible.

#### Closing libuv handles

If a libuv handle is not managed through a [`HandleWrap`][] instance,
it needs to be closed explicitly. Do not use `uv_close()` for that, but rather
`env->CloseHandle()`, which works the same way but keeps track of the number
of handles that are still closing.

#### Closing libuv requests

There is no way to abort libuv requests in general. If a libuv request is not
managed through a [`ReqWrap`][] instance, the

> `env->IncreaseWaitingRequestCounter()` and

Reviewer 5: I am slightly concerned direct references to these methods could get out of date. Can't we just write these down in env.h and link to the headers instead?

Author: > Can't we just write these down in env.h and link to the headers instead?

I’m not sure what you mean? But either way, I think keeping this doc up to date and making sure of that during reviews is something we’ll have to do either way.

Reviewer 5: > I think keeping this doc up to date and making sure of that during reviews is something we’ll have to do either way.

I think if keeping this doc up-to-date is going to be a blocker for landing changes in `src/`, we should just mov the information about the internal classes into the headers as comments, and link to those in the README. For one if the README is too long it becomes difficult to read, also it's going to be a pain to remember updating this doc while changing the code - but slightly less so if the comments live in the headers.

Also it feels weird to me that in order to understand some of these classes, you either have to read the code, or read this doc because what's supposed to be said in comments are moved into a markdown file...

Reviewer: tbh, I really like the proposed format. It has a flow, the necessary level of abstraction, example code at right places, and caveats and cautions as appropriate - something either not possible in code comments, or not easily consumable. IMO inline comments are rich with the context, but weak in connecting bigger views, and more importantly, discoverability.

Author: If you have items that you feel should also be documented more explicitly in a header file, I’m happy to do that, but generally I agree with @gireeshpunathil – this is not supposed to be an exhaustive guide or reference documentation, but rather paint the big picture and distinguish what’s important from what’s not.

Reviewer 5: > and more importantly, discoverability.

I feel the opposite - I think content in the README is more difficult to discover than content in the headers, especially when the README becomes long.

> this is not supposed to be an exhaustive guide or reference documentation, but rather paint the big picture and distinguish what’s important from what’s not.

That's what I think this document should be like as well, but with details like

> They can be added and removed through by using `env->AddCleanupHook(callback, hint);` and `env->RemoveCleanupHook(callback, hint);`, where callback takes a `void* hint` argument.

and 

>  If a libuv request is not managed through a [`ReqWrap`][] instance, the `env->IncreaseWaitingRequestCounter()` and `env->DecreaseWaitingRequestCounter()` functions need to be used to keep track of the number of active libuv requests.

It does not seem to be the case. These references to specific method names and their arguments do not seem to be necessary in terms of giving the reader a bigger view of the code base.

Author: > > and more importantly, discoverability.
> 
> I feel the opposite - I think content in the README is more difficult to discover than content in the headers, especially when the README becomes long.

That’s somewhat solvable by adding a table of contents, if you like?

> > this is not supposed to be an exhaustive guide or reference documentation, but rather paint the big picture and distinguish what’s important from what’s not.
> 
> That's what I think this document should be like as well, but with details like
> 
> > They can be added and removed through by using `env->AddCleanupHook(callback, hint);` and `env->RemoveCleanupHook(callback, hint);`, where callback takes a `void* hint` argument.
> 
> and
> 
> > If a libuv request is not managed through a [`ReqWrap`][] instance, the `env->IncreaseWaitingRequestCounter()` and `env->DecreaseWaitingRequestCounter()` functions need to be used to keep track of the number of active libuv requests.
> 
> It does not seem to be the case. These references to specific method names and their arguments do not seem to be necessary in terms of giving the reader a bigger view of the code base.

Well … where would you document these, especially the latter one? It’s easy to add inline docs to `ReqWrap` but what do you do when you want to describe when it’s about *not* using `ReqWrap`?

In any case, I don’t feel like these give away too much detail. Judging from the current contents of the file in this PR, the number of yearly PRs that would require updating it is probably in the single digits.

Reviewer 5: > That’s somewhat solvable by adding a table of contents, if you like?

I think that does not help in this case. To quote my edit:

> For me it's rare that I'll read a doc, top to bottom, to understand a code base and remember all these identifier names. I usually run into these names when I hack on the code and I expect to see explanations in the headers.

Now this is solvable if in the headers we point the readers to the README for more information, but that feels kind of weird.

> where would you document these, especially the latter one?

I think if the goal is to give a big picture in the README, for the latter one it can just say something like:

```
If a libuv request is not managed through a [ReqWrap][] instance, a counter is used to make sure it gets cleaned up.
```

And the method names can be documented in the `req_wrap.h`. The README can just point the readers to that header to see these details. This is a README for `src/` after all, so I'd expect it to be more of a table of contents for all these files under `src/`.

`env->DecreaseWaitingRequestCounter()` functions need to be used to keep track
of the number of active libuv requests.

#### Calling into JavaScript

Calling into JavaScript is not allowed during cleanup. Worker threads explicitly
forbid this during their shutdown sequence, but the main thread does not for
backwards compatibility reasons.

When calling into JavaScript without using [`MakeCallback()`][], check the
`env->can_call_into_js()` flag and do not proceed if it is set to `false`.

## Classes associated with JavaScript objects

### `MemoryRetainer`

A large number of classes in the Node.js C++ codebase refer to other objects.

> Inheriting from the `MemoryRetainer` class enables objects in a heap snapshot

Reviewer 5: The wording here is not quite clear about what `MemoryRetainer` does, I think - it's a helper for annotating C++ classes with information that can be used by the heap snapshot builder in V8, so that memory retained by C++ can be tracked in V8 heap snapshots captured in Node.js applications.

Also it's not exactly clear whether the `objects` here refer to JS objects or C++ objects?

Author: Thanks for the pointers, I’ve expanded this with wording based on your comment :+1:

get started working on it and document some idioms you may encounter while
doing so.

## Coding style

Node.js has a document detailing its [C++ coding style][]
that can be helpful as a reference for stylistic issues.

## V8 API documentation

A lot of the Node.js codebase is around what the underlying JavaScript engine,
V8, provides through its API for embedders. Knowledge of this API can also be
useful when working with native addons for Node.js written in C++, although for
new projects [N-API][] is typically the better alternative.

V8 does not provide much public API documentation beyond what is
available in its C++ header files, most importantly `v8.h`, which can be
accessed online in the following locations:

* On GitHub: [`v8.h` in Node.js master][]
* On GitHub: [`v8.h` in V8 master][]
* On the Chromium project’s Code Search application: [`v8.h` in Code Search][]

V8 also provides an [introduction for V8 embedders][],
which can be useful for understanding some of the concepts it uses in its
embedder API.

Important concepts when using V8 are the ones of [`Isolate`][]s and
[JavaScript value handles][].

> ## Libuv API documentation

Reviewer 6: ```suggestion
## libuv API documentation
```

functionalities to V8, most importantly the ability to schedule work on
background threads.

Node.js provides a `NodePlatform` class that implements the `v8::Platform`
interface and uses libuv for providing background threading abilities.

The platform can be accessed through `isolate_data->platform()` given an
[`IsolateData`][] instance, although that only works when:

* The current Node.js instance was not started by an embedder; or
* The current Node.js instance was started by an embedder whose `v8::Platform`
  implementation also implement’s the `node::MultiIsolatePlatform` interface
  and who passed this to Node.js.

<a id="binding-functions"></a>
### Binding functions

C++ functions exposed to JS follow a specific signature. The following example
is from `node_util.cc`:

```c++
void ArrayBufferViewHasBuffer(const FunctionCallbackInfo<Value>& args) {
  CHECK(args[0]->IsArrayBufferView());
  args.GetReturnValue().Set(args[0].As<ArrayBufferView>()->HasBuffer());
}
```

(Namespaces are usually omitted through the use of `using` statements in the
Node.js source code.)

`args[n]` is a `Local<Value>` that represents the n-th argument passed to the

> function. `args.This()` is the `this` value inside this function call.

Reviewer 6: `args.Holder()` is used below. Is it worth defining it here (or below)?

## Classes associated with JavaScript objects

### `MemoryRetainer`

A large number of classes in the Node.js C++ codebase refer to other objects.
The `MemoryRetainer` class is a helper for annotating C++ classes with
information that can be used by the heap snapshot builder in V8, so that
memory retained by C++ can be tracked in V8 heap snapshots captured in
Node.js applications.

Inheriting from the `MemoryRetainer` class enables objects (both from JavaScript
and C++) to refer to instances of that class, and in turn enables that class
to point to other objects as well, including native C++ types
such as `std::string` and track their memory usage.

This can be useful for debugging memory leaks.

The [`memory_retainer.h`][] header file explains how to use this class.

<a id="baseobject"></a>
### `BaseObject`

A frequently recurring situation is that a JavaScript object and a C++ object
need to be tied together. `BaseObject` is the main abstraction for that in
Node.js, and most classes that are associated with JavaScript objects are
subclasses of it. It is defined in [`base_object.h`][].

Every `BaseObject` is associated with one [`Environment`][] and one

> `v8::Object`. The `v8::Object` needs to have at least one “internal field”

Reviewer 6: It might be good to define internal fields somewhere.

overview over libuv handles managed by Node.js.

<a id="callback-scopes"></a>
### Callback scopes

The public `CallbackScope` and the internally use `InternalCallbackScope`
classes provide the same facilities as [`MakeCallback()`][], namely:

* Emitting the `'before'` event for async tracking when entering the scope
* Setting the current async IDs to the ones passed to the constructor
* Emitting the `'after'` event for async tracking when leaving the scope
* Running the `process.nextTick()` queue
* Running microtasks, in particular `Promise` callbacks and async/await
  functions

Usually, using `AsyncWrap::MakeCallback()` or using the constructor taking
an `AsyncWrap*` argument (i.e. used as
`InternalCallbackScope callback_scope(this);`) suffices inside of Node.js’s
C++ codebase.

## C++ utilities

Node.js uses a few custom C++ utilities, mostly defined in [`util.h`][].

### Memory allocation

Node.js provides `Malloc()`, `Realloc()` and `Calloc()` functions that work
like their C stdlib counterparts, but crash if memory cannot be allocated.
(As V8 does not handle out-of-memory situations gracefully, it does not make

> sense for us to attempt to do so in all cases.)

Reviewer 6: ```suggestion
sense for Node.js to attempt to do so in all cases.)
```

> # Node.js C++ codebase

Author: Yup, added it to the contributing guide (I don’t think there’s a good place inside the collaborator guide right now, it’s very focused on processes right now)

}
```

See [Callback scopes][] for more information.

<a id="handlewrap"></a>
### `HandleWrap`

`HandleWrap` is a subclass of `AsyncWrap` specifically designed to make working
with [libuv handles][] easier. It provides the `.ref()`, `.unref()` and
`.hasRef()` methods as well as `.close()` to enable easier lifetime management
from JavaScript. It is defined in [`handle_wrap.h`][].

`HandleWrap` instances are [cleaned up][cleanup hooks] automatically when the
current Node.js [`Environment`][] is destroyed, e.g. when a Worker thread stops.

`HandleWrap` also provides facilities for diagnostic tooling to get an
overview over libuv handles managed by Node.js.

<a id="reqwrap"></a>
### `ReqWrap`

`ReqWrap` is a subclass of `AsyncWrap` specifically designed to make working
with [libuv requests][] easier. It is defined in [`req_wrap.h`][].

In particular, its `Dispatch()` method is designed to avoid the need to keep
track of the current count of active libuv requests.

`ReqWrap` also provides facilities for diagnostic tooling to get an
overview over libuv handles managed by Node.js.

<a id="callback-scopes"></a>
### Callback scopes

> The public `CallbackScope` and the internally use `InternalCallbackScope`

Reviewer 7: ```suggestion
The public `CallbackScope` and the internally used `InternalCallbackScope`
```
?

  functions

Usually, using `AsyncWrap::MakeCallback()` or using the constructor taking
an `AsyncWrap*` argument (i.e. used as
`InternalCallbackScope callback_scope(this);`) suffices inside of Node.js’s
C++ codebase.

## C++ utilities

Node.js uses a few custom C++ utilities, mostly defined in [`util.h`][].

### Memory allocation

Node.js provides `Malloc()`, `Realloc()` and `Calloc()` functions that work
like their C stdlib counterparts, but crash if memory cannot be allocated.
(As V8 does not handle out-of-memory situations gracefully, it does not make
sense for Node.js to attempt to do so in all cases.)

The `UncheckedMalloc()`, `UncheckedRealloc()` and `UncheckedCalloc()` functions
return `nullptr` in these cases (or when `size == 0`).

#### Optional stack-based memory allocation

The `MaybeStackBuffer` class provides a way to allocate memory on the stack
if it is smaller than a given limit, and falls back to allocating it on the
heap if it is larger. This can be useful for performantly allocating temporary
data if it is typically expected to be small (e.g. file paths).

The `Utf8Value`, `TwoByteValue` (i.e. UTF-16 value) and `BufferValue`

> (`Utf8Value` but copy data from a `Buffer` is that is passed) helpers

Reviewer 8: Is there a typo here?

`DCHECK()`, `DCHECK_EQ()`, etc.

### Scope-based cleanup

The `OnScopeLeave()` function can be used to run a piece of code when leaving
the current C++ scope.

```c++
static void GetUserInfo(const FunctionCallbackInfo<Value>& args) {
  Environment* env = Environment::GetCurrent(args);
  uv_passwd_t pwd;
  // ...

  const int err = uv_os_get_passwd(&pwd);

  if (err) {
    // ... error handling, return early ...
  }

  auto free_passwd = OnScopeLeave([&]() { uv_os_free_passwd(&pwd); });

  // ...
  // Turn `pwd` into a JavaScript object now; whenever we return from this
  // function, `uv_os_free_passwd()` will be called.
  // ...
}
```

[`BaseObject`]: #baseobject
[`Context`]: #context
[`Environment`]: #environment
[`Global`]: #global-handles
[`HandleWrap`]: #handlewrap
[`IsolateData`]: #isolate-data
[`Isolate`]: #isolate
[`Local`]: #local-handles
[`MakeCallback()`]: #makecallback
[`MessagePort`]: https://nodejs.org/api/worker_threads.html#worker_threads_class_messageport
[`ReqWrap`]: #reqwrap
[`async_hooks` module]: https://nodejs.org/api/async_hooks.html
[`async_wrap.h`]: async_wrap.h
[`base_object.h`]: base_object.h
[`handle_wrap.h`]: handle_wrap.h

> [`memory_retainer.h`]: memory_retainer.h

Reviewer 8: Sorry to be late. This link seems 404 now.

[`req_wrap.h`]: req_wrap.h
[`util.h`]: util.h
[`v8.h` in Code Search]: https://cs.chromium.org/chromium/src/v8/include/v8.h
[`v8.h` in Node.js master]: https://github.com/nodejs/node/blob/master/deps/v8/include/v8.h
[`v8.h` in V8 master]: https://github.com/v8/v8/blob/master/include/v8.h
[`vm` module]: https://nodejs.org/api/vm.html
[C++ coding style]: ../CPP_STYLE_GUIDE.md
[Callback scopes]: #callback-scopes
[JavaScript value handles]: #js-handles
[N-API]: https://nodejs.org/api/n-api.html
[binding function]: #binding-functions
[cleanup hooks]: #cleanup-hooks
[event loop]: #event-loop
[exception handling]: #exception-handling

> [internal field]: #internal-field

Reviewer 8: ```diff
- [internal field]: #internal-field
+ [internal field]: #internal-fields
```

interface and uses libuv for providing background threading abilities.

The platform can be accessed through `isolate_data->platform()` given an
[`IsolateData`][] instance, although that only works when:

* The current Node.js instance was not started by an embedder; or
* The current Node.js instance was started by an embedder whose `v8::Platform`
  implementation also implement’s the `node::MultiIsolatePlatform` interface
  and who passed this to Node.js.

<a id="binding-functions"></a>
### Binding functions

C++ functions exposed to JS follow a specific signature. The following example
is from `node_util.cc`:

```c++
void ArrayBufferViewHasBuffer(const FunctionCallbackInfo<Value>& args) {
  CHECK(args[0]->IsArrayBufferView());
  args.GetReturnValue().Set(args[0].As<ArrayBufferView>()->HasBuffer());
}
```

(Namespaces are usually omitted through the use of `using` statements in the
Node.js source code.)

`args[n]` is a `Local<Value>` that represents the n-th argument passed to the
function. `args.This()` is the `this` value inside this function call,
and `args.Holder()` which is equivalent to `args.This()` in all use cases inside

> of Node.js.

Reviewer 6: I think this last sentence would make more sense if it was split into two sentences (suggestion below). Other than that, all of my previous nits look good 👍 

> `args.This()` is the `this` value inside this function call. `args.Holder()` is equivalent to `args.This()` in all use cases inside of Node.js.

Author: Sure, done!
